### PR TITLE
System tags CLI option

### LINKED ISSUE #485: Reduce cardinality of data sent to InfluxDB
K6 appears to generate very high cardinality data.

```
> show tag keys from "http_req_duration";
name: http_req_duration
tagKey
------
group
iter
method
name
ocsp_status
perf_type
proto
status
tls_version
url
vu
```

This is a lot of tags! And a few of them have very high cardinality (largely `iter`, but `vu` is not great either). This very quickly leads to a lot of series:

```
> show series from "http_req_duration" limit 20;
key
---
http_req_duration,group=::basic\ test\ 1,iter=0,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=1
http_req_duration,group=::basic\ test\ 1,iter=0,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=2
http_req_duration,group=::basic\ test\ 1,iter=0,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=3
http_req_duration,group=::basic\ test\ 1,iter=0,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=4
http_req_duration,group=::basic\ test\ 1,iter=0,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=5
http_req_duration,group=::basic\ test\ 1,iter=0,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=6
http_req_duration,group=::basic\ test\ 1,iter=0,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=7
http_req_duration,group=::basic\ test\ 1,iter=0,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=8
http_req_duration,group=::basic\ test\ 1,iter=0,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=9
http_req_duration,group=::basic\ test\ 1,iter=1,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=1
http_req_duration,group=::basic\ test\ 1,iter=1,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=2
http_req_duration,group=::basic\ test\ 1,iter=1,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=3
http_req_duration,group=::basic\ test\ 1,iter=1,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=4
http_req_duration,group=::basic\ test\ 1,iter=1,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=5
http_req_duration,group=::basic\ test\ 1,iter=1,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=6
http_req_duration,group=::basic\ test\ 1,iter=1,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=7
http_req_duration,group=::basic\ test\ 1,iter=1,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=8
http_req_duration,group=::basic\ test\ 1,iter=1,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=9
http_req_duration,group=::basic\ test\ 1,iter=10,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=1
http_req_duration,group=::basic\ test\ 1,iter=10,method=GET,name=Ex2-Basic1,proto=HTTP/1.1,status=200,url=http://test.loadimpact.com,vu=2
```

I was testing out k6 and very quickly hit my InfluxDB max series limit after about 10 minutes of benchmarking.

Currently I'm looking to use telegraf to exclude some of these fields, but it would be great if k6 supported choosing a few options:

1. Which of the default key-value pairs are fields vs tags.
2. Which default key-value pairs are used at all.
3. Which default metrics are used.

(1) alone will solve my actual issue, but the others would be nice :) Other than this issue have been very impressed with k6!

### PR DESCRIPTION
**Pull request by robingustafsson**
Adds a CLI option `--default-tags "url,method,status"` to specify which tags to include in metrics output, a whitelist. Collectors have the possibility to override and force certain tags to be included.

The following tags can be specified:
- `error` (http)
- `group` (http)
- `iter` (vu)
- `method` (http)
- `name` (http)
- `ocsp_status` (http)
- `proto` (http)
- `status` (http, websocket)
- `subprotocol` (websocket)
- `tls_version` (http)
- `url` (http, websocket)
- `vu` (vu)

Closes #485.

### CODE REVIEW

**Comment by robingustafsson** [comment]
@ppcano Could I remove `vu` and `iter` as tags now?

**Comment by ppcano** [comment]
@robingustafsson Yes, they can be removed. Confirmed by @artych

File: js/modules/k6/http/http_request.go [hunk 1]
```diff
@@ -137,16 +137,38 @@ func (h *HTTP) request(ctx context.Context, rt *goja.Runtime, state *common.Stat
...
-		"url":    url.URLString,
-		"name":   url.Name,
-		"group":  state.Group.Path,
-		"vu":     strconv.FormatInt(state.Vu, 10),
-		"iter":   strconv.FormatInt(state.Iteration, 10),
+	var tags map[string]string
+	if state.Options.DefaultTags == nil {
+		tags = map[string]string{
+			"proto":  "",
+			"status": "0",
```
Comment (na--) [inline js/modules/k6/http/http_request.go]: Why pre-populate the tags with wrong data? Shouldn't they only be written to when the actual request is made?

File: js/modules/k6/http/http_request.go [hunk 2]
```diff
@@ -137,16 +137,38 @@ func (h *HTTP) request(ctx context.Context, rt *goja.Runtime, state *common.Stat
...
+			"name":   url.Name,
+			"group":  state.Group.Path,
+		}
+	} else {
+		tags = map[string]string{}
+		if state.Options.DefaultTags["proto"] {
+			tags["proto"] = ""
+		}
+		if state.Options.DefaultTags["status"] {
+			tags["status"] = "0"
```
Comment (na--) [inline js/modules/k6/http/http_request.go]: Same question here for `proto` and `status`, and below for the websockets - why is this needed?

File: js/modules/k6/http/http_request.go [hunk 3]
```diff
@@ -334,7 +356,9 @@ func (h *HTTP) request(ctx context.Context, rt *goja.Runtime, state *common.Stat
 
 	if resErr != nil {
 		resp.Error = resErr.Error()
-		tags["error"] = resp.Error
+		if state.Options.DefaultTags == nil || state.Options.DefaultTags["error"] {
+			tags["error"] = resp.Error
+		}
```
Comment (na--) [inline js/modules/k6/http/http_request.go]: Shouldn't we actually update the `status` and `proto` fields here, even though there was an error. Arguably the `status` field is even more important in this case

Comment (robingustafsson) [reply js/modules/k6/http/http_request.go]: Agree, we should aim to set the tag values in connection with the context that the value reflects. So for `status` we should set it to 0 when handling an error, and `proto` should be set to an empty string.

File: js/modules/k6/http/http_request.go (hunk 1 above)
Comment (robingustafsson) [reply js/modules/k6/http/http_request.go]: Agree, see comment below.

File: js/modules/k6/http/http_request.go (hunk 2 above)
Comment (robingustafsson) [reply js/modules/k6/http/http_request.go]: Agree, see comment below.

File: js/modules/k6/http/http_request.go (hunk 3 above)
Comment (na--) [reply js/modules/k6/http/http_request.go]: Ah, I mistakenly assumed that non-nil `resErr` meant an HTTP error here, when it's mostly for things like network connectivity errors and such. So I'm not actually sure we should set `status` to 0 and `proto` to an empty string. Wouldn't it be better if we just don't include them at all, since 0 and an empty string are nonsensical values?

**Comment by na--** [comment]
@robingustafsson: something else is puzzling me, there are quite a lot of setters and getters for `lib.CollectorOptions`, but I don't see its `DefaultTags` actually being used anywhere. The `DefaultTags` (soon to be `SystemTags`) that are actually used are [the ones in](https://github.com/loadimpact/k6/pull/525/files#diff-2a3d79a0a11711742c3804c5e3846597R233) `lib.Options`...

Not sure if `lib.CollectorOptions` is a remnant of an earlier implementation or if it is supposed to be used in the future, but I don't think it's used now.

**Comment by robingustafsson** [comment]
@na-- True, we can remove the `lib.CollectorOptions` for now. The idea was that a collector could force certain tags to always be set. The use case we had was that the cloud collector needed the `vu` and `iter` tags for some result analysis algorithms in the cloud, but that dependency has been fixed/removed.

File: js/modules/k6/http/http_request.go (hunk 3 above)
Comment (robingustafsson) [reply js/modules/k6/http/http_request.go]: Yes, it's (mostly) for non/pre-HTTP errors. `status=0` is used in the cloud service in various places when querying the result data. We might come up with a better solution eventually, perhaps by only looking at the `error` tag, but not high priority right now. @ppcano Can answer better than me how reliant we currently are on `status=0` in the cloud service.

Comment (na--) [reply js/modules/k6/http/http_request.go]: Sorry, my latest push hid this comment section, hopefully commenting here will restore it, since my question is still valid with the new code. @ppcano, if this section still shows as outdated, please respond in the main PR comment thread.

**Comment by na--** [comment]
@ppcano, I have a few questions about the tags with regards to the cloud service and insights:
- (reiterating from the now hidden [inline discussion](https://github.com/loadimpact/k6/pull/525/files/b51a8126ea29bca640bff351dc4bf3e654beaa84#diff-37ac318df24f95fc3aadf765a5cfa593R361)) should we add a `status=0` tag for metrics with an error. I think that since they already have an `error` tag it shouldn't be necessary to send a nonsensical `status=0`, but if it's used in a lot of places, I can add it back
- if the users choose not to include `group` tag in the `--system-tags`, could it be safely excluded when uploading results to Insights or do we need to force it?

**Comment by ppcano** [comment]
@na-- 

> Should we add a status=0 tag for metrics with an error? 

Removing `status=0` is a breaking change in Insights. Right now, the cloud platform will have to add it when the `error` tag is present. I recommend to avoid this change at this moment.

>  not to include the group tag

`group` tag is a requirement for Insights, but it will work if `k6` force its inclusion when running `run -o cloud`.

**Comment by na--** [comment]
Ok, I'll add `status=0` back when there are errors. I'll also make sure that `group` is present when running `k6 cloud` or `k6 run -o cloud`. @ppcano, what's the full list of tags that are actually needed for LI Cloud/Insights? Are there any other besides`status` and `group`?

@robingustafsson: I considered bringing the `CollectorOptions` back so we can encode these requirements for the different collectors, but I think that in this case we don't have something optional, instead certain collectors have a requirement that certain tags be present in samples. So maybe adding a simple method like `GetRequiredSampleTags() Tags` to all collectors would be a better solution, since this is not something users should be able to change. 

Also, I don't think that silently adding tags back when the user has explicitly disabled them is the best way to do this. By default, if the user doesn't specify `--system-tags` manually, everything will work since the needed `cloud` collector tags will be included in the default option. But if the user manually disables the collection of `group`, `status`, etc. tags and tries to run or upload the results in the cloud, I think we should print an error rather than silently adding the disabled tags back. Doing otherwise would break the user's expectations, even if only in a small and probably insignificant way.

**Comment by robingustafsson** [comment]
@na-- Agree that error'ing out if the needed tags aren't set for the selected collector is a better approach. As you say it's probably going to be a rare event anyways. Adding a `GetRequiredSampleTags() Tags` (or `GetRequiredSystemTags()` perhaps?) to the `Collector` interface that can pass that information to the `run` and `cloud` command handlers feels like a good approach.

**Comment by ppcano** [comment]
@na-- 

> what's the full list of tags that are actually needed for LI Cloud/Insights? Are there any other besides `status` and `group`?

`name`, `method`, `status`, `error`, `check` and `group`.

@robingustafsson  I am not entirely convinced about making `name` and `status` optional.

Which will be the enabled tags by default?

**Comment by na--** [comment]
Right now I've got `"proto", "subproto", "status", "method", "url", "name", "group", "check", "error", "tls_version"` locally (still not pushed [in GitHub](https://github.com/loadimpact/k6/blob/feature/default-tags-option/lib/options.go#L36)). By default `iter`, `vu`, and `ocsp_status` won't be collected (am I missing something?). So pushing results to the cloud should work without issued by default. If the user disables any of the required tags you mentioned (incl. `name` and `status`) and tries to push results or execute in the cloud, I'll make it so a nice error message is printed.

**Comment by robingustafsson** [comment]
@na-- Talked to @ppcano, what you describe in your last comment sounds good.

**Comment by na--** [comment]
@robingustafsson @ppcano, I think that this is done, please have a look at the code and give it a :+1: if you approve. If it's fine, I'll update the docs tomorrow morning and merge this after. Considering all of the back-and-forth changes, I will probably have to squash it before merging :smile: 

The tests are currently failing because httpbin.org is very slow :roll_eyes:... I'll rerun them tomorrow morning again, hopefully by then it'll be OK. I think that my text task has to be removing all of the calls to it in the tests by using the new [replacement library](https://github.com/mccutchen/go-httpbin)...

File: js/modules/k6/ws/ws.go [hunk 4]
```diff
@@ -188,8 +189,12 @@ func (*WS) Connect(ctx context.Context, url string, args ...goja.Value) (*WSHTTP
 
 	defer func() { _ = conn.Close() }()
 
-	tags["status"] = strconv.Itoa(httpResponse.StatusCode)
-	tags["subprotocol"] = httpResponse.Header.Get("Sec-WebSocket-Protocol")
+	if state.Options.SystemTags["status"] {
+		tags["status"] = strconv.Itoa(httpResponse.StatusCode)
+	}
+	if state.Options.SystemTags["subproto"] {
+		tags["subproto"] = httpResponse.Header.Get("Sec-WebSocket-Protocol")
```
Comment (robingustafsson) [inline js/modules/k6/ws/ws.go]: Don't remember if I'm the one that made this change :), but if we change `subprotocol` to `subproto` I suspect the cloud service for example needs to be changed as well (@ppcano?).

**Comment by robingustafsson** [comment]
@na-- Besides the comment above regarding `subprotocol` to `subproto` the PR LGTM.

File: js/modules/k6/ws/ws.go (hunk 4 above)
Comment (na--) [reply js/modules/k6/ws/ws.go]: I'm the one who did it :smile:  I suggested it in an earlier comment in this PR and when nobody protested I assumed it was OK... :grinning: My rationale was that since we call the HTTP protocol tag `proto`, the subprotocol tag should be `subproto` for consistency. I can easily change it back if necessary.

Comment (ppcano) [reply js/modules/k6/ws/ws.go]: @na-- @robingustafsson 

> I suspect the cloud service for example needs to be changed as well 

Nothing has to be changed in the cloud service. The `subprotocol` tag is only used on the filter search and the tag filter works on any existing `tags` of the test results.

Comment (robingustafsson) [reply js/modules/k6/ws/ws.go]: @na-- 😄no worries, the name change makes sense 
@ppcano Ok, thought we had some special handling to show Websocket metrics in the breakdown?

Comment (ppcano) [reply js/modules/k6/ws/ws.go]: @robingustafsson 

To double check, I have reviewed the code, and we don't use the `subprotocol` tag for anything. The `ws` package emits different metrics (`ws_connecting`, `ws_session_duration` ....) and we check the metric name to handle it appropriately.

https://docs.k6.io/docs/k6-websocket-api

Comment (robingustafsson) [reply js/modules/k6/ws/ws.go]: @ppcano Ah, ok good.
@na-- I think the PR is good to merge then.

Comment (na--) [reply js/modules/k6/ws/ws.go]: Great, I'll merge it after I update the docs

**Comment by na--** [comment]
Updated https://docs.k6.io/v1.0/docs/tags-and-groups#section-system-tags and https://docs.k6.io/v1.0/docs/options, merging this...
